### DISCUSSION TITLE
[Feedback] Value/cost of `react-celo` vs off-the-shelf libraries

### DISCUSSION THREAD

**Discussion by arthurgousset** [celo-org/react-celo#352] [General]
## TLDR

Starting a discussion:
- on the value (if any) `react-celo` provides over the many off-the-shelf libraries that exist today (e.g. [Rainbowkit](https://github.com/celo-org/rainbowkit-celo), [Web3Modal](https://web3modal.com/), [Web3React](https://github.com/Uniswap/web3-react), [WAGMI](https://github.com/wagmi-dev/references/pull/79/files), [UseDapp](https://github.com/TrueFiEng/useDApp/pull/1079), etc)
- on the advantages/disadvantages of "_decoupling_" `react-celo` (separating the parts that are valuable and sunsetting the parts that are not)

## Context

@jmrossy's [Discord message](https://discord.com/channels/600834479145353243/783806028629934110/1079487782193463329):

> Would react-celo work without a contractkit peer dependency installed? The Mento UI is going to moving to a new custom SDK so it won't have any need for contractkit. It may make sense to keep using the react-celo modal + connection manager but not the web3.js + kit that come with it. My guess is something would break because I see a chain of deps from `react-celo -> @celo/wallet-* -> @celo/connect -> web3.js + peer ContractKit` deps. 
>
>If you think it would be a lot of work to decouple the modal UI from the kit, I think using a different off-the-shelf modal lib (e.g. RainbowKit, Web3Modal) would work fine for the Mento Dapp but the branding wouldn't be as 'Celo-ish'
>
> I would tag @aaronmgdr for his thoughts too but I see his name explicitly requests no tags haha. @jeanregisser also curious what you think since I see you made a `react-celo` change recently 🙂

@obernardovieira [Discord message](https://discord.com/channels/600834479145353243/783806028629934110/1079851113643257906):

> I've said the same before, a couple of times. From impactMarket side, we only care about modal and connection manager.
> I've considered changing it in the past, but didn't yet. Recently got back to the topic and we are planning to move.

**Comment by arthurgousset** [comment]
Hey @jmrossy, thanks for kicking off this conversation. 

### TLDR

- It appears that it would be a lot of work to decouple the modal UI from the kit.
- We'd recommend you use a different off-the-shelf modal lib
- We're looking into value/cost of maintaining `react-celo` in its entirety going forward (I'm not sold atm)

### More detail: 

We've been thinking about the value/cost of `react-celo` vs other off-the-shelf libraries actively since the new year.

I'm not sold on maintaining the entirety of `react-celo` going forward with my current understanding:
- Rainbowkit works well on Celo with the small plugin @aaronmgdr made and the team at cLabs like building with Rainbowkit:
  - [rainbowkit-with-celo.vercel.app](https://rainbowkit-with-celo.vercel.app/) 
- EthersjS works well on Celo since the [Blockchain client update last year](https://github.com/celo-org/celo-blockchain/pull/1957) (more in this convo)
  - https://github.com/celo-org/celo-monorepo/discussions/9922
- We're adding support for Celo in more and more off-the-shelf libraries
  - in WAGMI: https://github.com/wagmi-dev/references/pull/79
  - in UseDApp: https://github.com/TrueFiEng/useDApp/pull/1079
- The eng team is investigating ways to support custom gas fee currencies without wrapping entire libraries.
  - in EthersJS: https://github.com/ethers-io/ethers.js/issues/3747
  - in WalletConnect v2 (it appears like WCv2 doesn't strip custom fields anymore)
- The eng team is looking into moving the TS type interfaces out of contractkit into a smaller (standalone) package that can be used with your library of choice (Ethers/web3/etc).
  - https://github.com/celo-org/celo-monorepo/pull/9912

**Reply by jmrossy** [reply]
Thanks @0xarthurxyz for the response. I'm also not sold that continuing investment into this lib makes sense now that Celo works with more standard Eth tools.

I spoke with the Mento Labs team and we are also leaning towards replacing react-celo with wagmi + rainbowkit in the Mento UI. Now that you've confirmed that decoupling react-celo + contractkit would be difficult, it's pretty much a certainty that we'll go that route. 

Kudos to the fine work that @aaronmgdr @nicolasbrugneaux and others did to take react-celo as far as they did! It will be sad to lose the 'celo-native' feeling that this lib brought to celo dapps.

**Reply by arthurgousset** [reply]
> I spoke with the Mento Labs team and we are also leaning towards replacing react-celo with wagmi + rainbowkit in the Mento UI. 

Great, to know. If you folks run into any issues or frictions during the integration I'd love to know!

We're very keen to learn from you folks 🙏

**Reply by vibern0** [reply]
Thanks a lot @0xarthurxyz. Yeah, a bit sad losing that celo feeling, but in the end, I think only those starting on celo would use this lib. Anyone else coming from other networks, won't. And likely, even starting on celo, people might change to other libs. I'm not sold. And actually there's mainly one reason I did not migrate last time I was looking into it, which is an issue I'm having now again.

So, feedback on my first attempt to move. (rainbowkit + wagmi)
Note: I've done it first with impactMarket's utils (aka SDK) because, website, dapp and wallet depend on it, and there's an example app there.

It was fairly simple, I just have a blocker. Accepting cUSD as feeCurrency is a requiremnet.
wagmi has a `useSigner` which returns a signer, but it throws an error if I try to send a feeCurrency. Using celo-ethers-wrapper doesn't work from the dapp side (would work on the wallet I think). Changing celo-ethers-wrapper will also not work because I need a different signer, and that comes from wagmi. From dapp side I'm not sure how I can work around that. I see some are trying to fix it. Any feedback so far?
